### PR TITLE
Add output rules functionality with name, options and messages

### PR DESCRIPTION
**Pull request by RickvdStaaij**
### What

Output the rules of the validator and format them in a way you would like.

### How to test

1. See that unit tests have been added
1. See that the code coverage is 100% and code quality is 100%
1. See that all tests pass
1. See that you can output the rules in an array if no callable is given, and that you can format the resulted rules with a callable

### ToDo

* [ ] Include messages in the output
* [ ] Update documentation

### Issue

https://github.com/particle-php/Validator/issues/80

### CODE REVIEW

File: src/Chain.php [hunk 1]
```diff
@@ -337,6 +337,26 @@ public function validate(MessageStack $messageStack, Container $input, Container
...
+     */
+    public function getRules()
+    {
+        $rules = [];
+
+        /** @var Rule $rule */
+        foreach ($this->rules as $rule) {
+            $reflectionRule = new \ReflectionClass($rule);
+            $rules[] = [
+                'name' => $reflectionRule->getShortName(),
```
Comment (RickvdStaaij) [inline src/Chain.php]: Reflection might not be the best solution here, any suggestions on improving this? Adding a constant for every class containing the class name seems overkill to me.

Comment (berry-langerak) [inline src/Chain.php]: `substr(get_class($rule), strrpos(get_class($rule), NAMESPACE_SEPARATOR));` that'd yield the same result, I think.

**Comment by berry-langerak** [comment]
Hum. I had something else in mind, really. That's mostly because of the fact that I already thought out how this is supposed to work in my head, and I wasn't able to communicate that with you before falling of the end of the world -- sorry.

I'd rather see that there's *one* (and just one) method being added to Validator, which is `output`. Output can receive a callable, and to that callable we'll send a representation of the entire structure the validator has at that point. So a DTO containing other DTO's, pretty much. That way, we've decoupled the stuff that's being pushed to the callable from the actual structure within Validator, which is going to be helpful once we start refactoring (think of `each` support, for example).

That DTO should probably be passed along to every Chain the validator has, and to every Rule that a Chain has. They can then append their DTO representations of themselves, which means it's possible to keep the visibility of `getMessageParameters` to protected. Much cleaner, but more importantly: more future-proof. Would you mind terribly if I take this over?

**Comment by RickvdStaaij** [comment]
@berry-langerak No I wouldn't mind, but I would like to discuss this once you're back in the office, because I've been trying to put some effort into this, but I walked into some smaller things that I wanted to discuss with your before I was going to continue this. One of those is how the messages are currently stored in the validator, where I think this should be in an object with its own responsibility for it.

I put the WIP on this pretty much on hold until I could discuss this with you. Maybe we can arrange a skype call next week and make some plans on when and how.

**Comment by RickvdStaaij** [comment]
Closing in favor of #82

**Pull request closed without merging**